File: ios/SsdidWallet/Domain/Verifier/Offline/BundleFetcher.swift
import Foundation

/// Fetches DID documents from the registry and caches them as VerificationBundles
/// for offline credential verification. The TTL is determined by TtlProvider.
final class BundleFetcher {
    private let registryApi: RegistryApi
    private let bundleStore: BundleStore
    private let ttlProvider: TtlProvider

    init(registryApi: RegistryApi, bundleStore: BundleStore, ttlProvider: TtlProvider = TtlProvider()) {
        self.registryApi = registryApi
        self.bundleStore = bundleStore
        self.ttlProvider = ttlProvider
    }

    /// Resolve an issuer's DID document and cache it as a VerificationBundle.
    /// Returns nil on network errors (offline verification is best-effort).
    func fetchAndCache(issuerDid: String) async -> VerificationBundle? {
        do {
            let didDoc = try await registryApi.resolveDid(did: issuerDid)
            let formatter = ISO8601DateFormatter()
            let now = Date()
            let bundle = VerificationBundle(
                issuerDid: issuerDid,
                didDocument: didDoc,
                fetchedAt: formatter.string(from: now),
                expiresAt: formatter.string(from: now.addingTimeInterval(ttlProvider.ttl))
            )
            try await bundleStore.saveBundle(bundle)
            return bundle
        } catch {
            return nil // Network error — offline verification is best-effort
        }
    }
}

File: android/sdk/ssdid-core/src/main/java/my/ssdid/sdk/domain/verifier/offline/BundleFetcher.kt
package my.ssdid.sdk.domain.verifier.offline

import my.ssdid.sdk.domain.settings.TtlProvider
import my.ssdid.sdk.domain.transport.RegistryApi
import java.time.Instant

/**
 * Fetches DID documents from the registry and caches them as VerificationBundles
 * for offline credential verification. Bundle TTL is driven by TtlProvider.
 */
class BundleFetcher(
    private val registryApi: RegistryApi,
    private val bundleStore: BundleStore,
    private val ttlProvider: TtlProvider
) {
    /**
     * Resolve an issuer's DID document and cache it as a VerificationBundle.
     * Returns null on network errors (offline verification is best-effort).
     */
    suspend fun fetchAndCache(issuerDid: String): VerificationBundle? {
        return try {
            val didDoc = registryApi.resolveDid(issuerDid)
            val now = Instant.now()
            val bundle = VerificationBundle(
                issuerDid = issuerDid,
                didDocument = didDoc,
                fetchedAt = now.toString(),
                expiresAt = now.plus(ttlProvider.getTtl()).toString()
            )
            bundleStore.saveBundle(bundle)
            bundle
        } catch (_: Exception) {
            null // Network error — offline verification is best-effort
        }
    }
}
